From cee44531e619c94413e598ede38ab26b7bf653b2 Mon Sep 17 00:00:00 2001
From: Satoshi Ikeuchi <satosan.ikm@gmail.com>
Date: Mon, 19 Oct 2026 23:07:00 +0900
Subject: [PATCH] libjson-c: backport userdata cleanup fixes

Backport two upstream json-c fixes for object cleanup in version 0.19.
Clear the user-delete callback after calling it to avoid a second call
and double-free. Clear userdata so iterative cleanup cannot interpret
application data as a parent-object pointer.

Include the upstream regression test and its expected output. Bump the
package release to make the fixes available in updated packages.

Upstream: https://github.com/json-c/json-c/commit/3aa996b5b6bd96c29f2c6da98da33c3fb3d13038
Upstream: https://github.com/json-c/json-c/commit/0c3a5a1994ed6b5957d2e878fc1298ce9aa09b3f
Signed-off-by: Satoshi Ikeuchi <satosan.ikm@gmail.com>
(cherry picked from commit 102f4eebca24a5205aad14c183401d7799862e4d)

Refs: ADCDS/openwrt-xiaomi-ax3000t-rd03v2#50
---
 package/libs/libjson-c/Makefile               |  2 +-
 ...3-clear-user-delete-after-calling-it.patch | 98 +++++++++++++++++++
 ...clear-userdata-during-object-cleanup.patch | 48 +++++++++
 3 files changed, 147 insertions(+), 1 deletion(-)
 create mode 100644 package/libs/libjson-c/patches/003-clear-user-delete-after-calling-it.patch
 create mode 100644 package/libs/libjson-c/patches/004-clear-userdata-during-object-cleanup.patch

diff --git a/package/libs/libjson-c/Makefile b/package/libs/libjson-c/Makefile
index 4729adc62c..f707a0eb21 100644
--- a/package/libs/libjson-c/Makefile
+++ b/package/libs/libjson-c/Makefile
@@ -9,7 +9,7 @@ include $(TOPDIR)/rules.mk
 
 PKG_NAME:=json-c
 PKG_VERSION:=0.19
-PKG_RELEASE:=2
+PKG_RELEASE:=3
 
 PKG_SOURCE:=$(PKG_NAME)-$(PKG_VERSION)-nodoc.tar.gz
 PKG_SOURCE_URL:=https://github.com/json-c/json-c/releases/download/json-c-$(PKG_VERSION)-20260627
diff --git a/package/libs/libjson-c/patches/003-clear-user-delete-after-calling-it.patch b/package/libs/libjson-c/patches/003-clear-user-delete-after-calling-it.patch
new file mode 100644
index 0000000000..550294e16a
--- /dev/null
+++ b/package/libs/libjson-c/patches/003-clear-user-delete-after-calling-it.patch
@@ -0,0 +1,98 @@
+From 3aa996b5b6bd96c29f2c6da98da33c3fb3d13038 Mon Sep 17 00:00:00 2001
+From: Eric Hawicz <erh+git@nimenees.com>
+Date: Sat, 4 Jul 2026 11:02:06 -0400
+Origin: https://github.com/json-c/json-c/commit/3aa996b5b6bd96c29f2c6da98da33c3fb3d13038
+Subject: [PATCH] Issue #934: Be sure to clear _user_delete in
+ _json_object_put_maybe_free() to avoid calling it twice.  Add a test to
+ check.
+
+---
+ json_object.c             |  1 +
+ json_object.h             |  4 ++--
+ tests/test_deep_nesting.c | 33 +++++++++++++++++++++++++++++++++
+ 3 files changed, 36 insertions(+), 2 deletions(-)
+
+diff --git a/json_object.c b/json_object.c
+index 0a619679bd..6cb1bdac86 100644
+--- a/json_object.c
++++ b/json_object.c
+@@ -303,6 +303,7 @@ static inline int _json_object_put_maybe_free(struct json_object *jso, int free_
+ 
+ 	if (jso->_user_delete)
+ 		jso->_user_delete(jso, jso->_userdata);
++	jso->_user_delete = NULL;
+ 
+ 	switch (jso->o_type)
+ 	{
+diff --git a/json_object.h b/json_object.h
+index e21138ecf1..7edb63d4f8 100644
+--- a/json_object.h
++++ b/json_object.h
+@@ -259,7 +259,7 @@ JSON_EXPORT void *json_object_get_userdata(json_object *jso);
+  * The user_delete parameter is optional and may be passed as NULL, even if
+  * the userdata parameter is non-NULL.  It will be called just before the
+  * json_object is deleted, after it's reference count goes to zero
+- * (see json_object_put()).
++ * (see json_object_put()) but before any child objects are freed.
+  * If this is not provided, it is up to the caller to free the userdata at
+  * an appropriate time. (i.e. after the json_object is deleted)
+  *
+@@ -293,7 +293,7 @@ JSON_EXPORT void json_object_set_userdata(json_object *jso, void *userdata,
+  * The user_delete parameter is optional and may be passed as NULL, even if
+  * the userdata parameter is non-NULL.  It will be called just before the
+  * json_object is deleted, after it's reference count goes to zero
+- * (see json_object_put()).
++ * (see json_object_put()) but before any child objects are freed.
+  * If this is not provided, it is up to the caller to free the userdata at
+  * an appropriate time. (i.e. after the json_object is deleted)
+  *
+diff --git a/tests/test_deep_nesting.c b/tests/test_deep_nesting.c
+index 364c4820ea..9e251750c4 100644
+--- a/tests/test_deep_nesting.c
++++ b/tests/test_deep_nesting.c
+@@ -56,6 +56,37 @@ static void test_deep_nesting_tostring(const char *str)
+ 	json_tokener_free(tok);
+ }
+ 
++struct userdata_test {
++	int userdata_val;
++	char *p;
++};
++/*
++ * Check that the user_delete function is only called once, even with the
++ * newer code to avoid deeply nested calls during json_object_put().
++ */
++static void user_delete_test(struct json_object *jso, void *userdata_in)
++{
++	struct userdata_test *userdata = (struct userdata_test *)userdata_in;
++	printf("in user_delete, userdata_val=%d\n", userdata->userdata_val);
++	fflush(stdout);
++	userdata->userdata_val = 0;
++	userdata->p[0] = 'x';
++	userdata->p[8191] = 'x';
++	free(userdata->p);
++}
++static void test_nesting_with_user_delete(void)
++{
++	json_object *jso;
++	struct userdata_test userdata_val = {
++		1, malloc(8192)
++	};
++
++ 	jso = json_object_new_object();
++	json_object_set_userdata(jso, &userdata_val, user_delete_test);
++	json_object_object_add(jso, "somekey", json_object_new_string("foo"));
++	json_object_put(jso);
++}
++
+ int main(int argc, char **argv)
+ {
+ 	char *str;	
+@@ -77,5 +108,7 @@ int main(int argc, char **argv)
+ 
+ 	free(str);
+ 
++	test_nesting_with_user_delete();
++
+ 	return EXIT_SUCCESS;
+ }
diff --git a/package/libs/libjson-c/patches/004-clear-userdata-during-object-cleanup.patch b/package/libs/libjson-c/patches/004-clear-userdata-during-object-cleanup.patch
new file mode 100644
index 0000000000..8698b3eb5e
--- /dev/null
+++ b/package/libs/libjson-c/patches/004-clear-userdata-during-object-cleanup.patch
@@ -0,0 +1,48 @@
+From 0c3a5a1994ed6b5957d2e878fc1298ce9aa09b3f Mon Sep 17 00:00:00 2001
+From: Eric Hawicz <erh+git@nimenees.com>
+Date: Sat, 4 Jul 2026 11:38:15 -0400
+Origin: https://github.com/json-c/json-c/commit/0c3a5a1994ed6b5957d2e878fc1298ce9aa09b3f
+Subject: [PATCH] Issue #934: Also set _userdata=NULL in
+ _json_object_maybe_free() so json_object_put doesn't try to use it as a
+ _delete_parent value. Update the test_deep_nesting test output.
+
+---
+ json_object.c                    | 7 +++++--
+ tests/test_deep_nesting.expected | 1 +
+ 2 files changed, 6 insertions(+), 2 deletions(-)
+
+diff --git a/json_object.c b/json_object.c
+index 71cd25b2a6..095b84e2b4 100644
+--- a/json_object.c
++++ b/json_object.c
+@@ -304,6 +304,8 @@ static inline int _json_object_put_maybe_free(struct json_object *jso, int free_
+ 	if (jso->_user_delete)
+ 		jso->_user_delete(jso, jso->_userdata);
+ 	jso->_user_delete = NULL;
++	jso->_userdata = NULL; // aka _delete_parent, but json_object_put() will
++	                       // have already grabbed it if it needs it.
+ 
+ 	switch (jso->o_type)
+ 	{
+@@ -435,10 +437,11 @@ int json_object_put(struct json_object *jso)
+ 
+ 		// All slots are cleared, now pop back up to the parent
+ 		{
+-			json_object *parent = jso->_delete_parent;
+-			int rc;
+ 			// jso is a child that's already been detached from its parent
+ 			// so we need to actually free it now
++			// Be sure to grab _delete_parent *before* freeing jso.
++			json_object *parent = jso->_delete_parent;
++			int rc;
+ 			assert(jso->_ref_count == 0);
+ 			jso->_ref_count++;   // We're the exclusive owner of jso, non-atomic add is ok.
+ 			// Note: the call must not be inside assert(), or it gets
+diff --git a/tests/test_deep_nesting.expected b/tests/test_deep_nesting.expected
+index b869cfaabe..b9c1ab9848 100644
+--- a/tests/test_deep_nesting.expected
++++ b/tests/test_deep_nesting.expected
+@@ -1,2 +1,3 @@
+ Parsed depth 100000 string to json_object: yes
+ Freed json_object
++in user_delete, userdata_val=1
-- 
2.47.3

